cpp/src/io/avro/avro.cpp

```diff
@@ -92,36 +94,113 @@ bool container::parse(file_metadata* md, size_t max_num_rows, size_t first_row)
       }
     }
   }
+  // Save the first sync markers in the metadata; we compare them to other
+  // sync markers that should be present at the end of a block.  If they
+  // differ, the data should be interpreted as corrupted.
   md->sync_marker[0] = get_raw<uint64_t>();
   md->sync_marker[1] = get_raw<uint64_t>();
 
   md->metadata_size  = m_cur - m_base;
-  md->skip_rows      = 0;
+  md->skip_rows      = first_row;
+  md->total_num_rows = 0;
   max_block_size     = 0;
   total_object_count = 0;
+  // Enumerate the blocks in this file.  Each block starts with a count of
+  // objects (rows) in the block (uint64_t), and then the total size in bytes
+  // of the block (uint64_t).  We walk each block and do the following:
+  //    1. Capture the total number of rows present across all blocks.
+  //    2. Add each block to the metadata's list of blocks.
+  //    3. Handle the case where we've been asked to skip or limit rows.
+  //    4. Verify sync markers at the end of each block.
+  //
+  // A row offset is also maintained, and added to each block.  This reflects
+  // the absolute offset that needs to be added to any given row in order to
+  // get the row's index within the destination array.  See `dst_row` in
+  // `avro_decode_row()` for more information.
+  //
+  // N.B. "object" and "row" are used interchangeably here; "object" is
+  //      avro nomenclature, "row" is ours.
+  //
+  // N.B. If we're skipping rows, we ignore blocks (i.e. don't add them to
+  //      md->block_list) that precede the block containing the first row
+  //      we're interested in.
+  //
+  // N.B. The 18 below is (presumably) intended to account for the two 64-bit
+  //      object count and block size integers (16 bytes total), and then an
+  //      additional two bytes to represent the smallest possible row size.
   while (m_cur + 18 < m_end && total_object_count < max_num_rows) {
     auto const object_count = static_cast<uint32_t>(get_encoded<int64_t>());
     auto const block_size   = static_cast<uint32_t>(get_encoded<int64_t>());
-    if (block_size <= 0 || object_count <= 0 || m_cur + block_size + 16 > m_end) { break; }
-    if (object_count > first_row) {
-      auto block_row = static_cast<uint32_t>(total_object_count);
+    auto const next_end     = m_cur + block_size + 16;
+    // Abort on terminal conditions.  We keep these as separate lines instead of
+    // combining them into a single if in order to facilitate setting specific
+    // line breakpoints in the debugger.
+    if (block_size <= 0) { return false; }
+    if (object_count <= 0) { return false; }
+    if (next_end > m_end) { return false; }
```

Comment:
CUDF_EXPECTS?
We can leave the existing exists as they are, but it would be nice to start switching to exceptions in the new code.
Edit: the change is already a huge improvement :+1: